batchcloser/closers.go

```diff
@@ -0,0 +1,111 @@
+package batchcloser
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/reddit/baseplate.go/batcherror"
+)
+
+// CloseError is used to wrap the errors returned by the inner closers within a
+// BatchCloser.
+//
+// It can be used to inspect both the error that was returned and the io.Closer
+// that caused it.
+type CloseError struct {
+	Cause  error
+	Closer io.Closer
+}
+
+// Error implements the interface for error.
+func (err CloseError) Error() string {
+	return fmt.Sprintf("batchcloser: error closing closer %#v : %s", err.Closer, err.Cause.Error())
+}
+
+// Unwrap implements helper interface for errors.Is.
+func (err CloseError) Unwrap() error {
+	return err.Cause
+}
+
+// As implements helper interface for errors.As.
+func (err CloseError) As(v interface{}) bool {
+	if target, ok := v.(*CloseError); ok {
+		*target = err
+		return true
+	}
+	if target, ok := v.(**CloseError); ok {
+		*target = &err
+		return true
+	}
+	if errors.As(err.Cause, v) {
+		return true
+	}
+	return false
+}
+
+type simpleCloser struct {
+	close func() error
+}
+
+func (c simpleCloser) Close() error {
+	return c.close()
+}
+
+// Wrap can be used wrap a function that has the same signature as io.Closer.Close
```

Comment:
🔕 I still find the saying of "same signature as io.Closer.Close" a bit weird.

This is a lambda of `func() error`, that by itself should be self-describing enough. Also the good thing about lambdas is that they don't really care whatever your original signature is, it's always easy to wrap into a lambda. Say your original signature is actually `func(a, b int)`, you can still wrap into that by:

```go
// You must already have a & b here, of course.
batchcloser.Wrap(func() error {
  myFunc(a, b)
  return nil
})
```